### DISCUSSION TITLE
JPALazyDataModel iterator performance

### DISCUSSION THREAD

**Discussion by FlipWarthog** [primefaces/community#2014] [PrimeFaces] [answered]
In our implementation of JPALazyDataModel we have methods to return the full lists of both filtered and unfiltered values using `iterator()` + `IteratorUtils.toList` which are primarily used for custom CSV exports. We found that this will run N queries to assemble all the data (N = # rows / # pages), which changes based on rows per page selection, and causes certain operations to timeout for taking too long.

I made a recent performance enhancement to these to instead call `load` for the full set of data in one shot.

```diff
   // getLazyDataModel() returns the current instance of JPALazyDataModel
   public List<T> getValues() {
-      return IteratorUtils.toList(
-               getLazyDataModel().iterator());
+      final LazyDataModel<T> model = getLazyDataModel();
+      Map<String, SortMeta> sortMeta = this.getBinding().getSortByAsMap();
+      Map<String, FilterMeta> filterMeta = Collections.emptyMap();
+      return model.load(0, model.count(filterMeta), sortMeta, filterMeta);
   }

   public List<T> getFilteredValues() {
-      return IteratorUtils.toList(
-               getLazyDataModel().iterator(this.getBinding().getSortByAsMap(), this.getBinding().getFilterByAsMap()));
+      final LazyDataModel<T> model = getLazyDataModel();
+      Map<String, SortMeta> sortMeta = this.getBinding().getSortByAsMap();
+      Map<String, FilterMeta> filterMeta = this.getBinding().getFilterByAsMap();
+      return model.load(0, model.count(filterMeta), sortMeta, filterMeta);
   }
```

This is mostly FYI + WDYT? Should we consider changing the impl of `iterator` for JPALazyDataModel to take this efficiency into consideration? @Rapster @tandraschko @melloware

**Comment by melloware** [comment]
i think this seems reasonable to me.

**Comment by tandraschko** [comment, marked as answer]
i think your method shouldnt use the iterator

the iterator is for lazy loop through pages, not a findAll to get all objects into memory even if your loop may stop after 10 objects

**Reply by FlipWarthog** [reply]
That's my thought too; we're technically going against the intention of Iterator by loading it all into a List for how we use it, so I'll mark this as "Answered" and it'll be here for other people's awareness depending on how they use it.

**Reply by tandraschko** [reply]
it would be great if you can add some javadocs

**Reply by FlipWarthog** [reply]
Since the `iterator` impls (below) are actually in LazyDataModel, there's not a good spot for javadocs, I don't think. Would it make more sense as an added note in datatable.md under the `JPALazyDataModel` section? I can type something up and submit a PR for it.

```java
    @Override
    public Iterator<T> iterator() {
        return new LazyDataModelIterator<>(this);
    }

    public Iterator<T> iterator(Map<String, SortMeta> sortBy, Map<String, FilterMeta> filterBy) {
        return new LazyDataModelIterator<>(this, sortBy, filterBy);
    }
```

**Reply by FlipWarthog** [reply]
Let me know what you think.
https://github.com/primefaces/primefaces/pull/12003
